### DISCUSSION TITLE
Playback & Recording with wav format in the same program

### DISCUSSION THREAD

**Discussion by massa007** [pschatzmann/arduino-audio-tools#1009] [Q&A] [answered]
Hello audio-tools community :) 

Thanks a lot to pschatzmann for developing and helping out people with your free software, it is highly appreciated!!

I am currently playing with audio tools for simple playing and recording of files. I am now struggling with bringing the raw files into more comfortable wav format. I saw the example code but I am just too unexperienced to put playback and recording together in the same program. I just attached the important parts of my code which are working fine while using raw outputs.

Can anyone lead me to the right direction on how to playback and record something with wav format?

**preambel:**
```
uint16_t sample_rate = 16000;
uint64_t end_time; // trigger to call endRecord
uint8_t channels = 1;  // The stream will have 2 channels 

AudioKitStream kit;
File file;   // final output stream
StreamCopy copier; // copies data
```

**functions:**
```

 void record_start(){
   Serial.println("Recording...");
   Serial.println(file_name);
   // open the output file
   file.close();
   file = SD.open(file_name, FILE_WRITE);
   copier.begin(file, kit);   
   recording=true;  
 }
 
 void playback_start(){   // bool pinStatus, int pin, void* ref
   file.close();
   Serial.println("Play File...");
   file = SD.open(file, FILE_READ);
   copier.begin(kit, file);  
   playback = true; 
 }
```

**setup:**

```
 auto cfg = kit.defaultConfig(RXTX_MODE);
   cfg.sd_active = true;
   cfg.sample_rate = sample_rate;
   cfg.channels = channels;
   cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
   kit.begin(cfg);
   kit.setVolume(1.0);
```

**loop:**
`copier.copy();`

**Comment by pschatzmann** [comment, marked as answer]
Just read the chapter about encoding and decoding in the Wiki!
